nemo/collections/nlp/models/nlp_model.py

```diff
@@ -312,101 +241,238 @@ def setup(self, stage: str) -> None:
+                    checkpoint_version = json.load(f).get('checkpoint_version', None)
+                # get checkpoint version
+                logging.info(
+                    (
+                        f'Detected model parallel .nemo file: {restore_path}. '
+                        f'Assuming megatron model parallelism with '
+                        f'model_parallel_size: {app_state.model_parallel_size} '
+                        f'and checkpoint version: {checkpoint_version}'
+                    )
+                )
+            os.chdir(cwd)
+
         if app_state.model_parallel_size is not None:
-            # filepath needs to be updated to include mp_rank
-            dirname = os.path.dirname(filepath)
-            basename = os.path.basename(filepath)
-            filepath = f'{dirname}/mp_rank_{app_state.model_parallel_rank:02d}/{basename}'
+            if not isinstance(trainer, Trainer):
+                raise ValueError("trainer must be a PyTorch Lightning Trainer to restore model parallel .nemo files.")
 
-            # dump states as a checkpoint dictionary object
-            checkpoint = self.dump_checkpoint(weights_only)
+            if checkpoint_version is None:
+                raise ValueError(
+                    "Restoring from megatron model parallel .nemo but could not find megatron checkpoint version."
+                )
+            else:
+                logging.info(f"Setting megatron checkpoint version: {checkpoint_version}")
+                set_checkpoint_version(checkpoint_version)
 
-            # each model parallel rank needs to save a copy of its model
-            if app_state.data_parallel_rank == 0:
-                # write the checkpoint dictionary on the file
-                if self.trainer.accelerator_backend:
-                    checkpoint = self.trainer.accelerator_backend.on_save(checkpoint)
-                try:
-                    atomic_save(checkpoint, filepath)
-                except AttributeError as err:
-                    if LightningModule.CHECKPOINT_HYPER_PARAMS_KEY in checkpoint:
-                        del checkpoint[LightningModule.CHECKPOINT_HYPER_PARAMS_KEY]
-                    rank_zero_warn(
-                        'Warning, `hyper_parameters` dropped from checkpoint.' f' An attribute is not picklable {err}'
-                    )
-                    atomic_save(checkpoint, filepath)
-        return None
+            app_state.world_size = trainer.num_gpus * trainer.num_nodes
+
+            # try to get local rank from global
+            local_rank = None
+            try:
+                local_rank = int(os.environ['LOCAL_RANK'])
+            except:
+                logging.info('Global variable LOCAL_RANK not yet specified. Assuming LOCAL_RANK is 0.')
+
+            if local_rank is not None:
+                app_state.local_rank = local_rank
+            else:
+                # if local is None then we are on the main process
+                local_rank = 0
```

Comment:
This assumption is risky - on other envs all of the nodes will assume they are local rank 0 and load that checkpoint multiple times. Everything works but inference would give random results.

Best to simply raise a hard error here.

Reply:
agree with @titu1994 here, let's raise an exception here.